.github/workflows/ci.yml

```diff
@@ -0,0 +1,290 @@
+name: ci
+on:
+  pull_request:
+  push:
+    branches:
+      - master
+defaults:
+  run:
+    shell: bash
+jobs:
+  check-lint:
+    runs-on: ubuntu-22.04
+    steps:
+      - name: ci/checkout-repo
+        uses: actions/checkout@ac593985615ec2ede58e132d2e21d2b1cbd6127c # v3.3.0
+      - name: ci/setup-node
+        uses: actions/setup-node@64ed1c7eab4cce3362f8c340dee64e5eaeef8f7c # v3.6.0
+        id: setup_node
+        with:
+          node-version-file: ".nvmrc"
+          cache: npm
+      - uses: actions/cache@2b8105bdae4b746666ee225970c4a281bbd9d51f # v3.2.4
```

Comment:
Caching speeds up things in order not to download/install node modules again since it is a costly operation . This is also up for discussion

Reply:
If cache hits, then `npm install` will not run. 

https://github.com/mattermost/mattermost-webapp/pull/12182/files#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR30-R31

There is logic injected, that makes sures that node_modules and node_dependencies step will be skipped in the event of cache hit. 
```
if: steps.npm-cache.outputs.cache-hit != 'true'
```

Caching node modules, and node dependencies is considered as best practise, since it can save CPU cycles/network calls, thus reduce the CI execution time.

Reply:
Ah, I missed the cache-hit part.

Unfortunately, this has some other issues since, on a cache hit, it won't cache the compiled subpackages which are symlinked from `node_modules` since those are generated as part of `make node_modules`. Those live in:
- `packages/client/lib`
- `packages/types/lib`
- `packages/components/dist`
and we'll be adding more of those over time as the web app is split into more subpackages.

Does GHA have any equivalent to the CircleCI workspaces I started using in https://github.com/mattermost/mattermost-webapp/pull/11648 where we don't need to specify explicit file paths but can still share copies of the working directory between CI jobs?

Reply:
There is `upload-artifact` action, that can be used to pass data between jobs. 
https://docs.github.com/en/actions/migrating-to-github-actions/migrating-from-circleci-to-github-actions#persisting-data-between-jobs

But it is not recommended for the whole repository. It is designed for small quantity of artifacts, and not whole repositories. 

To unblock this, let me know if we are aligned removing the Caching part completely. And revisit at a later stage. Without node-modules caching, the pipeline finishes in about 14minutes (including benchmarks), therefore I feel that we hit the performance target even without caching.

Reply:
Caching disabled for node_modules with: https://github.com/mattermost/mattermost-webapp/commit/1cb8e2afb7b9fa69b45c605d1ea32a32310ea046

Reply:
Sounds good. I think that's about as long as CircleCI+benchmarks took before as well